### ISSUE TITLE
Concerning "Module A (Import B : S)" notation

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#1434]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 1434
Date: 2007-03-08 16:59:10 +0100
From: emakarov@lix.polytechnique.fr
To:   

Last updated: 2007-03-08 16:59:10 +0100

## Bugzilla Comment ID: 1511
Date: 2007-03-08 16:59:10 +0100
From: emakarov@lix.polytechnique.fr

Full_Name: Yevgeniy Makarov
Version: Trunk Revision: 9679
OS: Linux
Submission from: sphinx.lix.polytechnique.fr (129.104.11.1)

This concerns the notation "Module A (Import B : S)", which declares module B as
an argument and imports it at the same time. This notation is not described in
the section 2.5 "Module system" in the reference manual, but it is mentioned in
the CHANGES file for revision 8927.

Consider the following declaration of two module types:

Module Type A.

Parameter N : Set.
Parameter plus : N -> N -> N.

End A.

Module Type B.

Parameter N : Set.
Parameter mult : N -> N -> N.

End B.

(The plus and mult parameters intentionally overlap with the corresponding
definitions in Coq.Init.Peano.) I would expect that the following line:

Module C (Import MA : A) (Import MB : B).

is equivalent to

Module C (MA : A) (MB : B).
Import MA.
Import MB.

However, in the first case above, "Locate plus" command prints

Constant MA.plus
Constant Coq.Init.Peano.plus
  (shorter name to refer to it in current context is Peano.plus)

(as expected) and "Locate mult" command prints

Constant Coq.Init.Peano.mult
Constant MB.mult (shorter name to refer to it in current context is MB.mult)

i.e., the module MB does not seem to be imported, since the default
interpretation of mult comes from Coq.Init.Peano.

In the second variant above, both MA and MB are imported, so the default
interpretations of plus and mult come from these modules.

To sum up, the keyword Import in the parameter list is not a mere syntactic
sugar since in the case of two module parameters it is not equivalent to the
expected expansion.

## Bugzilla Comment ID: 1512
Date: 2007-07-04 17:04:28 +0200
From: emakarov@lix.polytechnique.fr

Corrigé (r9693)
